### PR TITLE
Reactive routes - add @Param, @Header and @Body

### LINKED ISSUE #10596: Allow injecting parameters and body in reactive routes
**Description**
The idea behind this improvement would be to support the injection of request parameters and body as argument of the reactive route method:

```java
@Route(path="/")
public void (RoutingContext rc, @Param("name") String name, Person body)
```

Parameter injection need to be annotated with a specific annotation (maybe reuse the jax-rs one) to indicate which parameter need to be injected.
Optionaly might be a good idea too (`Optional<String>`). Type conversion also. Multiple parameter (List) might also be a nice plus.

For the body, we should support Buffer, JsonObject/JsonArray and object (mapped from body using `bodyAsBuffer().toJson().mapTo(Person.class)`
I don't believe a specific annotation is required for the body. But I may be wrong.

### PR DESCRIPTION
**Pull request by mkouba**
- refactor param injectors
- resolves #10596

### CODE REVIEW

**Comment by mkouba** [comment]
Damn, I used `mvn clean install -Dquickly` and hence formatting errors..

**Comment by cescoffier** [comment]
Is path parameter supported? Should it uses the same `@Param` annotation or a specific one?

**Review by cescoffier — APPROVED** [review]
LGTM! I'm wondering about path parameter.

**Comment by gastaldi** [comment]
It looks like multiple parameters are not supported (eg. http://localhost:8080/foo?choice=1&choice=2).
Same for Header parameters : https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

**Comment by mkouba** [comment]
> Are path parameters supported? Should it use the same @Param annotation or a specific one?

@cescoffier They are supported. `HttpServerRequest.getParam()` returns both the path params and query params. I believe that it should use the same annotation.

> It looks like multiple parameters are not supported

@gastaldi Yes, we always return the first value. But it should be easy to implement injection of `List<String>`. I'm on it...

**Comment by mkouba** [comment]
@gastaldi Done ;-).

File: docs/src/main/asciidoc/reactive-routes.adoc [hunk 1]
```diff
@@ -193,10 +196,70 @@ The annotated method can accept arguments of the following types:
 * `io.vertx.reactivex.core.http.HttpServerRequest`
 * `io.vertx.reactivex.core.http.HttpServerResponse`
 
-If the annotated method returns `void` then it has to accept at least one argument. 
-If the annotated method does not return `void` then the arguments are optional.
+Furhtermore, it is possible to inject the `HttpServerRequest` parameters into a method parameter annotated with `@io.quarkus.vertx.web.Param`:
```
Comment (gastaldi) [inline docs/src/main/asciidoc/reactive-routes.adoc:199]:
```suggestion
Furthermore, it is possible to inject the `HttpServerRequest` parameters into a method parameter annotated with `@io.quarkus.vertx.web.Param`:
```

File: docs/src/main/asciidoc/reactive-routes.adoc [hunk 2]
```diff
@@ -193,10 +196,70 @@ The annotated method can accept arguments of the following types:
...
+|`java.lang.String`   |`routingContext.request().getParam()`   
+|`java.util.Optional<String>`   |`routingContext.request().getParam()`   
+|`java.util.List<String>`   |`routingContext.request().params().getAll()`
+|===
+
+.Request Parameter Example
+[source,java]
+----
+@Route
+String hello(@Param Optional<String> name) {
```
Comment (gastaldi) [inline docs/src/main/asciidoc/reactive-routes.adoc:214]: Isn't a compilation flag required to read attribute names?

File: docs/src/main/asciidoc/reactive-routes.adoc (hunk 1 above)
Comment (mkouba) [reply docs/src/main/asciidoc/reactive-routes.adoc:199]: Thanks for spotting this!

File: docs/src/main/asciidoc/reactive-routes.adoc (hunk 2 above)
Comment (mkouba) [reply docs/src/main/asciidoc/reactive-routes.adoc:214]: You're right. We should improve the error message...

Comment (mkouba) [reply docs/src/main/asciidoc/reactive-routes.adoc:214]: FYI: https://github.com/quarkusio/quarkus/issues/4632#issuecomment-657407680

**Review by gastaldi — APPROVED** [review]
LGTM, good job!

**Pull request merged**